### PR TITLE
updated filters. added pub sub node

### PR DESCRIPTION
**Pull request by shubhankar12**
Moved all the filters to a new directory. Added node to subscribe raw images and pub processed images. Tested on a sample video.

### CODE REVIEW

File: src/preprocess/preprocess_node.cpp [hunk 1]
```diff
@@ -6,47 +6,47 @@
 #include <cv_bridge/cv_bridge.h>
 #include <sensor_msgs/image_encodings.h>
 
-#include "whitebalance.h"
-#include "median_subtract.h"
+#include "filters/whitebalance.h"
+#include "filters/median_subtract.h"
+#include "filters/bilateral_filter.h"
+
+#define IMAGE_INPUT_RAW "/camera/image_raw"
```
Comment (abrandemuehl) [inline src/preprocess/preprocess_node.cpp]: Lets try and take these as parameters instead of hardcoded

File: src/preprocess/preprocess_test.cpp [hunk 2]
```diff
@@ -0,0 +1,53 @@
+#include <ros/ros.h>
```
Comment (abrandemuehl) [inline src/preprocess/preprocess_test.cpp]: Maybe we should actually turn this into a real set of tests

File: launch/preprocess.launch [hunk 3]
```diff
@@ -0,0 +1,5 @@
+<launch>
+  <param name="IMAGE_INPUT_RAW" value="/camera/image_raw" type="str" />
```
Comment (abrandemuehl) [inline launch/preprocess.launch]: Lets not capitalize all our parameters. They should look the same as the topic names

File: src/preprocess/preprocess_node.cpp [hunk 4]
```diff
@@ -1,51 +1,63 @@
...
+#include "filters/whitebalance.h"
+
+image_transport::Publisher filter_img_pub;
 
 using namespace cv;
+namespace enc = sensor_msgs::image_encodings;
+
+void imageCallback(const sensor_msgs::ImageConstPtr& msg) {
+  cv_bridge::CvImageConstPtr cv_ptr;
+  // obatianing the image from the camera topic
```
Comment (abrandemuehl) [inline src/preprocess/preprocess_node.cpp]: Typo

File: src/preprocess/preprocess_node.cpp [hunk 5]
```diff
@@ -1,51 +1,63 @@
...
+  try {
+    cv_ptr = cv_bridge::toCvShare(msg, enc::BGR8);
+  } catch (cv_bridge::Exception& e) {
+    ROS_ERROR("Could not convert from '%s' to 'bgr8'.", msg->encoding.c_str());
```
Comment (abrandemuehl) [inline src/preprocess/preprocess_node.cpp]: The program won't exit at this point, which it should

File: src/preprocess/preprocess_node.cpp [hunk 6]
```diff
@@ -1,51 +1,63 @@
...
-  Mat image;
-  image = imread(imageName.c_str(), IMREAD_COLOR);
+  std::string input_image_topic;
+  std::string output_image_topic;
 
-  if (image.empty()) {
-    std::cout << "Could not open or find the image" << std::endl;
-    return -1;
+  if (!ros::param::get("IMAGE_INPUT_RAW", input_image_topic)) {
+    throw ros::Exception("Must specify angular_position_topic parameter");
```
Comment (abrandemuehl) [inline src/preprocess/preprocess_node.cpp]: Is this the right parameter?

File: src/preprocess/preprocess_node.cpp [hunk 7]
```diff
@@ -1,51 +1,63 @@
...
   }
+  if (!ros::param::get("IMAGE_OUTPUT_FILTERED", output_image_topic)) {
+    throw ros::Exception("Must specify angular_position_topic parameter");
```
Comment (abrandemuehl) [inline src/preprocess/preprocess_node.cpp]: Is this the right error message?

File: launch/preprocess.launch [hunk 8]
```diff
@@ -0,0 +1,5 @@
+<launch>
+  <param name="/raw_img_topic" value="/camera/image_raw" type="str" />
```
Comment (abrandemuehl) [inline launch/preprocess.launch]: Make these parameters private to the node

```
<node ...>
 <param .../>
</node>
```

File: src/preprocess/CMakeLists.txt [hunk 9]
```diff
@@ -1,3 +1,8 @@
-add_executable(${PROJECT_NAME}_node preprocess_node.cpp whitebalance.cpp median_subtract.cpp)
-add_dependencies(${PROJECT_NAME}_node ${${PROJECT_NAME}_EXPORTED_TARGETS} ${catkin_EXPORTED_TARGETS})
-target_link_libraries(${PROJECT_NAME}_node ${catkin_LIBRARIES} ${OpenCV_LIBRARIES})
+add_executable(preprocess_node preprocess_node.cpp filters/bilateral_filter.cpp filters/whitebalance.cpp filters/median_subtract.cpp)
```
Comment (abrandemuehl) [inline src/preprocess/CMakeLists.txt]: Wait for #47 to go through so you can use the cmake macros for all of this

File: src/preprocess/filters/bilateral_filter.h [hunk 10]
```diff
@@ -0,0 +1,4 @@
+#pragma once
```
Comment (abrandemuehl) [inline src/preprocess/filters/bilateral_filter.h]: For consistency, this should be an ifdef guard

File: src/preprocess/preprocess_test.cpp [hunk 11]
```diff
@@ -0,0 +1,52 @@
+#include <cv_bridge/cv_bridge.h>
```
Comment (abrandemuehl) [inline src/preprocess/preprocess_test.cpp]: We can probably just remove this file

Comment (abrandemuehl) [reply src/preprocess/preprocess_test.cpp]: Maybe we could create a launch file that displays the image after processing

**Comment by abrandemuehl** [comment]
I'm rewriting a bunch of this stuff. Should be simpler now

**Comment by abrandemuehl** [comment]
@shubhankar12 What do you think of the changes I just made? I think it simplifies it a bit

**Comment by shubhankar12** [comment]
Looks good to me. This pull request has been open for a while.
@abrandemuehl

File: src/preprocess/preprocess_node.cpp [hunk 12]
```diff
@@ -1,51 +1,50 @@
...
-  Mat image;
-  image = imread(imageName.c_str(), IMREAD_COLOR);
+  ros::NodeHandle nh;
+  image_transport::ImageTransport it(nh);
 
-  if (image.empty()) {
-    std::cout << "Could not open or find the image" << std::endl;
-    return -1;
-  }
+  sub = it.subscribe("input", 1, callback);
```
Comment (shubhankar12) [inline src/preprocess/preprocess_node.cpp]: read topics from launch files

Comment (abrandemuehl) [reply src/preprocess/preprocess_node.cpp]: I'm having the launch file do remapping instead. I think it's less error prone and more standard

Comment (shubhankar12) [reply src/preprocess/preprocess_node.cpp]: oh ok. that also works.

**Comment by abrandemuehl** [comment]
I removed the travis stuff

**Comment by abrandemuehl** [comment]
Does it look good? If so I'll approve the changes and merge

**Comment by shubhankar12** [comment]
Yes, looks good. 
@abrandemuehl

**Pull request merged**